lib/Frontend/Frontend.cpp

```diff
@@ -261,233 +262,300 @@ ModuleDecl *CompilerInstance::getMainModule() {
-  if (!implicitHeaderPath.empty()) {
-    if (!clangImporter->importBridgingHeader(implicitHeaderPath, MainModule)) {
-      importedHeaderModule = clangImporter->getImportedHeaderModule();
-      assert(importedHeaderModule);
-    }
-  }
+  ModuleDecl *importedHeaderModule = importBridgingHeader(clangImporter);
 
   SmallVector<ModuleDecl *, 4> importModules;
-  if (!options.ImplicitImportModuleNames.empty()) {
-    for (auto &ImplicitImportModuleName : options.ImplicitImportModuleNames) {
-      if (Lexer::isIdentifier(ImplicitImportModuleName)) {
-        auto moduleID = Context->getIdentifier(ImplicitImportModuleName);
-        ModuleDecl *importModule = Context->getModule(std::make_pair(moduleID,
-                                                                 SourceLoc()));
-        if (importModule) {
-          importModules.push_back(importModule);
-        } else {
-          Diagnostics.diagnose(SourceLoc(), diag::sema_no_import,
-                               ImplicitImportModuleName);
-          if (Invocation.getSearchPathOptions().SDKPath.empty() &&
-              llvm::Triple(llvm::sys::getProcessTriple()).isMacOSX()) {
-            Diagnostics.diagnose(SourceLoc(), diag::sema_no_import_no_sdk);
-            Diagnostics.diagnose(SourceLoc(),
-                                 diag::sema_no_import_no_sdk_xcrun);
-          }
-        }
-      } else {
-        Diagnostics.diagnose(SourceLoc(), diag::error_bad_module_name,
-                             ImplicitImportModuleName, false);
-      }
-    }
+  fillInModulesToImportFromImplicitImportModuleNames(importModules);
+
+  if (Invocation.getInputKind() == InputFileKind::IFK_Swift_REPL) {
+    supplyREPLFileWithImports(implicitModuleImportKind,
+                              objCModuleUnderlyingMixedFramework,
+                              importedHeaderModule, importModules);
+    return;
   }
 
-  auto addAdditionalInitialImports = [&](SourceFile *SF) {
-    if (!underlying && !importedHeaderModule && importModules.empty())
-      return;
+  std::unique_ptr<DelayedParsingCallbacks> DelayedCB(
+      computeDelayedParsingCallback());
 
-    using ImportPair =
-        std::pair<ModuleDecl::ImportedModule, SourceFile::ImportOptions>;
-    SmallVector<ImportPair, 4> additionalImports;
-
-    if (underlying)
-      additionalImports.push_back({ { /*accessPath=*/{}, underlying },
-                                    SourceFile::ImportFlags::Exported });
-    if (importedHeaderModule)
-      additionalImports.push_back({ { /*accessPath=*/{}, importedHeaderModule },
-                                    SourceFile::ImportFlags::Exported });
-    if (!importModules.empty()) {
-      for (auto &importModule : importModules) {
-        additionalImports.push_back({ { /*accessPath=*/{}, importModule }, {} });
-      }
-    }
+  PersistentParserState PersistentState;
 
-    SF->addImports(additionalImports);
-  };
+  ensureMainFileComesFirst(implicitModuleImportKind,
```

Comment:
"Ensure" is the wrong name here – that would imply checking or modifying some existing list of files. I think it's just "createMainFile".

Reply:
Also, this tuple of implicit import info is coming up a lot, possibly enough to factor into a struct.

Reply:
Great catch! I'm doing it. When you see what I've done, I'm sure you can come up with better names, too.